File: src/leetcode/763partitionLabels.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    vector<int> partitionLabels(string s) {
        int n = s.size();
        vector<int> ans;
        // 记录每个字符最后出现的位置
        vector<int> last(26, 0);
        for (int i = 0; i < n; i++) {
            last[s[i] - 'a'] = i;
        }
        int left = 0, right = 0;
        for (int i = 0; i < n; i++) {
            // 更新当前分区的最远边界
            right = max(right, last[s[i] - 'a']);
            if (i == right) {
                // 到达当前分区的最远边界，添加分区长度
                ans.push_back(right - left + 1);
                left = i + 1;
            }
        }
        return ans;
    }
};


int main(int argc, char *argv[])
{
    string s = "ababcbacadefegdehijhklij";
    vector<int> ans = Solution().partitionLabels(s);
    for (int i = 0; i < ans.size(); i++){
        cout << ans[i] << endl;
    }
    return 0;
}

File: scripts/leetcode/763partitionLabels.py
'''
解题思路：贪心
1. 注意为什么使用ord(ch) - ord("a"), 而不能直接使用ch - "a"
'''
from typing import List
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        last = [0] * 26
        for i, ch in enumerate(s):
            last[ord(ch) - ord("a")] = i
        left = 0
        right = 0
        ans = list()
        for j, ch in enumerate(s):
            right = max(right, last[ord(ch) - ord("a")])
            if j == right:
                ans.append(right - left + 1)
                left = right + 1
        return ans

s = "ababcbacadefegdehijhklij"
print(f"result: {Solution().partitionLabels(s)}")